compiler/src/iree/compiler/Codegen/Common/ConvertBf16ArithToF32.cpp

```diff
@@ -0,0 +1,312 @@
+      for (auto attr : op->getAttrs()) {
+        auto newAttr = convertAttribute(op->getLoc(), attr.getValue(),
+                                        *getTypeConverter());
+        newAttrs.push_back(NamedAttribute(attr.getName(), newAttr));
+      }
+    } else {
+      newAttrs.append(op->getAttrs().begin(), op->getAttrs().end());
+    }
+
+    llvm::SmallVector<Type, 4> newResults;
+    (void)getTypeConverter()->convertTypes(op->getResultTypes(), newResults);
+
+    OperationState state(op->getLoc(), op->getName().getStringRef(), operands,
+                         newResults, newAttrs, op->getSuccessors());
+
+    for (Region &r : op->getRegions()) {
+      Region *newRegion = state.addRegion();
+      rewriter.inlineRegionBefore(r, *newRegion, newRegion->begin());
+      TypeConverter::SignatureConversion result(newRegion->getNumArguments());
+      (void)getTypeConverter()->convertSignatureArgs(
+          newRegion->getArgumentTypes(), result);
+      rewriter.applySignatureConversion(newRegion, result);
+    }
+
+    Operation *newOp = rewriter.create(state);
+    rewriter.replaceOp(op, newOp->getResults());
+    return success();
+  }
+};
+
+template <typename OpTy, typename TypeTy,
+          typename OperandToResultWidthLegalityRelation>
+struct ConvertTypeSensitiveArithCastOp : public OpConversionPattern<OpTy> {
+  using OpConversionPattern<OpTy>::OpConversionPattern;
+  LogicalResult matchAndRewrite(
+      OpTy op, typename OpTy::Adaptor adaptor,
+      ConversionPatternRewriter &rewriter) const override {
+    auto resultType =
+        this->getTypeConverter()->convertType(op.getResult().getType());
+    auto operandType =
+        this->getTypeConverter()->convertType(op.getOperand().getType());
+
+    auto resultEType = cast<TypeTy>(getElementTypeOrSelf(resultType));
+    auto operandEType = cast<TypeTy>(getElementTypeOrSelf(operandType));
+    // If post-conversion, the types would be equal, then the op becomes a
+    // no-op. Note that the op does not itself allow such a configuration, so we
+    // have to catch this before creating the new op.
+    if (resultEType == operandEType) {
+      rewriter.replaceOp(op, adaptor.getOperands()[0]);
+      return success();
+    }
+    // If after conversion the op becomes invalid, but not same-type (which we
+    // can fold above), then bail out.
+    // TODO: In some cases, we can repair the situation here, but for integer
+    // truncation, we don't know whether we should invert with signed or
+    // unsigned extension.
+    if (!OperandToResultWidthLegalityRelation()(operandEType.getWidth(),
+                                                resultEType.getWidth())) {
+      return rewriter.notifyMatchFailure(op, "invalid width combination");
+    }
+    rewriter.replaceOpWithNewOp<OpTy>(op, resultType, op.getOperand());
+    return success();
+  }
+};
+
+template <typename SrcOp>
+class PropagateCastF : public OpRewritePattern<SrcOp> {
```

Comment:
I think it would be useful to add some comments as to why this is needed. I understood based on offline discussion. It wasnt clear to me from the code that this is for handling some weird gap in arith op + vector type issues.

Reply:
Do we have an issue for this issue?

Reply:
updated the comment and included the bug.